Add APIs to set/get replication config (#1314)
minio/minio-go: docs/API.md

| `opts.SendContentMd5`          | _bool_                 | Specify if you'd like to send `content-md5` header with PutObject operation. Note that setting this flag will cause higher memory usage because of in-memory `md5sum` calculation. |
| `opts.PartSize`                | _uint64_               | Specify a custom part size used for uploading the object                                                                                                                           |
| `opts.ReplicationVersionID`                | _string_               | Specify VersionID of object to replicate.This option is intended for internal use by MinIO server to extend the replication API implementation by AWS. This option should not be set unless the application is aware of intended use.                                                                                              |

> | `opts.ReplicationStatus`                | _minio.ReplicationStatus_ | Specify replication status of object. This option is intended for internal use by MinIO server to extend the replication API implementation by AWS. This option should not be set unless the application is aware of intended use.                                                                                                             |

Reviewer: ```suggestion
| `opts.ReplicationStatus`                | _minio.ReplicationStatus_ | Set replication status of object. This option is intended for internal use by MinIO server to extend the replication API implementation by AWS. This option should not be set unless the application is aware of intended use.                                                                                                             |
```

> | `opts.ReplicationMTime`                | _time.Time_               | Specify replication modTime of source to be used for object. This option is intended for internal use by MinIO server to extend the replication API implementation by AWS. This option should not be set unless the application is aware of intended use.                                                                                                             |

Reviewer: ```suggestion
| `opts.ReplicationMTime`                | _time.Time_               | Preserve source modTime on the replicated object. This option is intended for internal use only by MinIO server to comply with AWS bucket replication implementation. This option should not be set unless the application is aware of intended use.                                                                                                             |
```

| `opts.PartSize`                | _uint64_               | Specify a custom part size used for uploading the object                                                                                                                           |

> | `opts.ReplicationVersionID`                | _string_               | Specify VersionID of object to replicate.This option is intended for internal use by MinIO server to extend the replication API implementation by AWS. This option should not be set unless the application is aware of intended use.                                                                                              |

Reviewer: We should validate that on the client that this is `uuid` - since this is MinIO specific extension.

Author: added a validation here.

> | `opts.ReplicationStatus`                | _string_               | Specify replication status of object. This option is intended for internal use by MinIO server to extend the replication API implementation by AWS. This option should not be set unless the application is aware of intended use.                                                                                                             |

Reviewer: can we have a type for ReplicationStatus?

> | `opts.ReplicationMTime`                | _string_               | Specify replication modTime of source to be used for object. This option is intended for internal use by MinIO server to extend the replication API implementation by AWS. This option should not be set unless the application is aware of intended use.                                                                                                             |

Reviewer: This should be time.Time @poornas not string

Author: done

fmt.Printf("%+v\n", versioningConfig)
```

<a name="SetBucketReplication"></a>

### SetBucketReplication(ctx context.Context, bucketname, cfg replication.Config, replicationARN string) error
Set replication configuration on a bucket. Prior to setting replication configuration, replication endpoint should have been defined on MinIO
using `mc admin bucket replication set` to associate the bucket `bucketName` with the replication endpoint. Next, this call requires a replication ARN which can be fetched from MinIO server using `mc admin bucket remote`

__Parameters__

|Param   |Type   |Description   |
|:---|:---| :---|
|`ctx`  | _context.Context_  | Custom context for timeout/cancellation of the call|
|`bucketName` | _string_  |Name of the bucket|
|`cfg` | _replication.Config_  |Replication configuration to be set |

> |`replicationARN` | _string_  |Replication ARN from MinIO server identifying the replication endpoint |

Reviewer: I think personally we should extend the

```
<ReplicationConfiguration>
   <Role>string</Role>
   <ReplicationArn></ReplicationArn> --> optional for AWS
   <Rule>
      <DeleteMarkerReplication>
         <Status>Disabled</Status>
      </DeleteMarkerReplication>
      <Destination>
         <Bucket>string</Bucket>
         <StorageClass>string</StorageClass>
      </Destination>
      <Filter>
         <And>
            <Prefix>string</Prefix>
            <Tag>
               <Key>string</Key>
               <Value>string</Value>
            </Tag>
            ...
         </And>
         <Prefix>string</Prefix>
         <Tag>
            <Key>string</Key>
            <Value>string</Value>
         </Tag>
      </Filter>
      <ID>string</ID>
      <Prefix>string</Prefix>
      <Priority>integer</Priority>
      <Status>string</Status>
   </Rule>
</ReplicationConfiguration>
```

Author: this is a good idea, makes it cleaner. Done

      </DeleteMarkerReplication>
      <Destination>
         <Bucket>string</Bucket>
         <StorageClass>string</StorageClass>
      </Destination>
      <Filter>
         <And>
            <Prefix>string</Prefix>
            <Tag>
               <Key>string</Key>
               <Value>string</Value>
            </Tag>
            ...
         </And>
         <Prefix>string</Prefix>
         <Tag>
            <Key>string</Key>
            <Value>string</Value>
         </Tag>
      </Filter>
      <ID>string</ID>
      <Prefix>string</Prefix>
      <Priority>integer</Priority>
      <Status>string</Status>
   </Rule>
</ReplicationConfiguration>`
replicationConfig := replication.Config{}
if err := xml.Unmarshal([]byte(replicationStr), &replicationConfig); err != nil {
    log.Fatalln(err)
}
replicationArn := "arn:minio:s3::598361bf-3cec-49a7-b529-ce870a34d759:*"
err = minioClient.SetBucketReplication(context.Background(), "my-bucketname", replicationConfig, replicationArn)
if err != nil {
    fmt.Println(err)
    return
}
```

<a name="GetBucketReplication"></a>

### GetBucketReplication(ctx context.Context, bucketName string) (replication.Config, error)

> Get replication config on a bucket.

Reviewer: ```suggestion
Get current replication config on a bucket.
```
